### ISSUE TITLE
Formal AppImage specification

### ISSUE THREAD

**Issue by shoogle** [probonopd/AppImageKit#161]
**TLDR** - I think it would be a good idea to have a formal AppImage specification because it would:

1. Be a single, definitive source of information about the AppImage format.
2. Ensure AppImages behave in a consistent manner, thereby enabling third-party developers to write AppImage management applications and distributions to offer built-in support for AppImages.
3. Serve as a guide for AppImageKit development.
4. Provide a developer who with a checklist for creating AppImages.
5. Enable a lintian-style application to be written to check AppImages for compliance with the standard.
6. Indicate the direction of the project and warn developers of future changes to the AppImage format.
7. Cement the open nature of the AppImage format.

--------------------------

@probonopd, I know you have been reluctant to specify things formally in the past for two reasons:

1. The format could change.
2. We should be flexible and allow developers to do things differently if they want. (We don't want to create another "distribution".)

However, users and developers appreciate it when programs work together in a consistent manner, and I think that the lack of a formal specification is actually preventing AppImages from gainer wider adoption. Even a "draft" specification would be extremely valuable, and it would still be valuable even if many of its components were optional rather than compulsory because it would help to reduce fragmentation of the AppImage concept.

It would be best to avoid a situation where every application developer adopts a different solution to the same set of problems. There are some problems that AppImages have yet to solve completely, but by offering draft specifications and "requests for comments" developers would get a clear picture of what they can expect in the future and how they should prepare for it.

Writing a specification is much quicker and easier than writing the software to implement it, so the draft specification could be used to formalise feature that are not yet included in AppImageKit and to give code contributors a guide as to where their efforts would be most appreciated. Features would be "draft" or "proposed" until implemented within AppImageKit, at which point they would become "official".

Documentation for AppImages is currently spread across a number of locations, including AppImage.org, the AppImageKit README and AppImages README and the Wiki. There are also some things that are only documented in the code, like the [fact that the remote update URL is stored in the [ISO 9660 Volume Descriptor field](https://github.com/probonopd/AppImages/blob/6b582cb8535b381f7ddb5fa816af2a827d797f5c/bintray.sh#L182-L196). I think that it would be helpful to have a complete and definitive description of the AppImage format in one place.

There are many formats offering similar features to AppImages. Not all of these formats are as open or as platform-agnostic as the AppImage format, and it would be a tragedy if a closed format became the new de facto standard for application packaging. To prevent this from happening, we need to make it absolutely clear what the AppImage format offers and which direction it is heading in.

---------------------------------------------

### How the specification could work:
Many standards-based projects use GitHub to host specifications (e.g. [The OpenAPI Specification](https://github.com/OAI/OpenAPI-Specification/blob/OpenAPI.next/versions/3.0.md)). The specification would be hosted in a dedicated Git repository (perhaps called AppImageSpec?). The repository could have a structure like this:

File | Purpose
---|---
README.md | Instructions for contributing
LICENSE | All-important copyright license
draft.md | Working document with proposed changes for the next specification release
latest.md | The latest version of the specification
versions/ | Directory for previous versions of the specification
versions/v1.0.md | Version 1 of the specification
versions/v2.0.md | Version 2... etc.

Proposals would be discussed via the issue tracker and then submitted via pull request for final approval by @probonopd. The specification would be kept as concise as possible and would not include any kind of explanation, justification or help with implementation (although it could link to that information). The specification itself would only include the "what", and not the "why" or "how". The detailed justifications and explanation could be kept in a parallel annotated version of the specification, or by linking to the discussion in the actual pull request. Help documentation would be maintained in a public Wiki.

AppImageKit would be a reference implementation of building standards-compliant AppImages, and the AppImages repository would provide examples. 

---------------------------------------------

#### 1. Embedding a remote update URL
You **MAY** choose to embed a URL in the AppImage's ISO 9660 Volume Descriptor field. If you do this, the URL **MUST** always point to the latest release of the AppImage (i.e. not a specific version). The URL **MUST** point directly to an actual AppImage file, **NOT** a webpage.

#### 2. Handling environment variables
Environment Variable | Effect on AppImage behaviour
---|---
LD_LIBRARY_PATH | AppImages **SHOULD** ignore this and use their internal libraries.
APPIMAGE_LIBRARY_PATH | **OPTIONAL** override to force linking against external libraries.

#### 3. Command line options
You **MAY** choose to implement any number of AppImage-specific command line options in addition to the ordinary command line options available in non-AppImage version of your program. Some examples of possible commands and their behaviours are listed below. If implemented, the command **MUST** result in the behaviour specified. The command **MUST** be available after a `--AppImage` option is specified, though you **MAY** choose to make it work without the `--AppImage` prefix if this does not conflict with any of your program's native commands.

Option | Behaviour if implemented (**MUST** be non-interactive unless otherwise stated)
---|---
options | Display a list of AppImage supported specific options and their behaviours.
man | Display the program's UNIX manual page (man page).
install | Perform desktop integration for the current user (or all users if run as root).
remove | Remove desktop integration files for the current user (all users if run as root).
-i | Enable interactivity for any of the other options

**Usage:**
```bash
./Program.AppImage --AppImage <option>  # Perform option non-interactively
./Program.AppImage --AppImage -i <option>  # Perform option interactively
```

**Example:**
```bash
./Program.AppImage --AppImage install  # Move AppImage to ~/bin and copy desktop integration files to ~/.local/share
./Program.AppImage --AppImage -i <option>  # Ask user where to install (default locations as above)
sudo ./Program.AppImage --AppImage install  # Move AppImage to /usr/local and copy desktop integration files to /usr/local/share
```

**Comment by probonopd** [comment]
Thanks @shoogle for pushing this.
In addition to what you outlined above, I think we should layer the spec:
* Layer 1: The filesystem format and main entry point (e.g., ISO9660 with ELF, RockRidge, zisofs, etc.)
* Layer 2: What is inside the AppImage (AppDir, what the AppRun file actually does, usr/, .desktop file, icons, metadata...) - this _might_ contain many best practices/guidelines
* Layer 3: Interfaces to the outside (e.g. command line parameters and environment variables AppImages might react to)

What I am least certain about is what should be hard requirements rather than loose conventions. The beauty of AppImage is that essentially anyone can do what s/he wants with it while it still works. You can make an AppImage that does everything slightly differently and the AppImage might still work technically (the same would not be true for, say a rpm or deb because the package manager would simply refuse to work correctly). The downside of this "you can do anything" approach is that AppImages are not predictable, and there are few 3rd-party tools that interact with AppImages so far ([Firejail](https://github.com/netblue30/firejail#appimage) being one recent exception). So probably some more formalization will indeed help to enable an ecosystem around the AppImage format. I guess that we need to start discussing these things rather than to avoid them being discussed simply because they might change at some point.

In other words, I am looking forward to a lively discussion on the "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" and would lean toward more of the latter unless there is good reason for the former.

So with that, I have set up https://github.com/AppImage/Spec and am looking forward to issues and pull requests.

**Comment by shoogle** [comment]
@probonopd many thanks for pursuing this! Sorry to be a pain, but I would highly recommend changing the repository name from "Spec" to something with "AppImage" in the title (maybe "AppImageSpec"). I know the repository belongs to the "AppImage" account on GitHub so it's full title is "AppImage/Spec", but as soon as I (or anyone else) creates a fork of it then that will no longer be true (i.e. it will be called "Shoogle/Spec"). I think it would be better to preserve the "AppImage" branding.

**Comment by probonopd** [comment]
You are right.
